### PR TITLE
Customer io

### LINKED ISSUE #4185: [ACTION] Customer.io Add People to Segment
**Is there a specific app this action is for?**
Customer.io
**Please provide a link to the relevant API docs for the specific service / operation.**
Add customers to segment https://www.customer.io/docs/api/#operation/add_to_segment

### PR DESCRIPTION
**Pull request by vellames**
Closes #4185

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4492"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

File: components/customer_io/actions/add-customers-to-segment/add-customers-to-segment.mjs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    },
+    idType: {
+      label: "Id Type",
+      type: "string",
+      description: "The type of ids you want to use. All of the values in the `ids` array must be of this type. If you don't provide this parameter, we assume that the `ids` array contains `id` values.",
+      options: [
+        "id",
+        "email",
+        "cio_id",
+      ],
```
Comment (lcaresia) [inline components/customer_io/actions/add-customers-to-segment/add-customers-to-segment.mjs:32]: Move this to a `constants` file.

```suggestion
      options: constants.ID_TYPES,
```

File: components/customer_io/actions/create-or-update-customer/create-or-update-customer.mjs [hunk 2]
```diff
@@ -1,60 +1,44 @@
...
-      },
-    };
-
-    return axios($, config);
+    const res = await this.app.createOrUpdateCustomer(this.customerId, {
+      email: this.email,
+      created_at: this.created_at,
+      ...this.attributes,
+    }, $);
+    $.export("$summary", "Successfully created/updated customer");
```
Comment (lcaresia) [inline components/customer_io/actions/create-or-update-customer/create-or-update-customer.mjs:41]: Obs: O didn't tested this.

```suggestion
    $.export("$summary", `Successfully created/updated customer with ID ${this.customerId}`);
```

File: components/customer_io/actions/send-event-to-customer-io/send-event-to-customer-io.mjs [hunk 3]
```diff
@@ -1,60 +1,45 @@
...
-      },
+    const data = {
+      name: this.eventName,
+      type: this.type,
+      data: this.customData,
     };
-
-    return axios($, config);
+    const res = await this.app.sendEventTo(this.customerId, data, $);
+    $.export("$summary", "Successfully sent event");
```
Comment (lcaresia) [inline components/customer_io/actions/send-event-to-customer-io/send-event-to-customer-io.mjs:42]: Obs: I didn't tested this.

```suggestion
    $.export("$summary", `Successfully sent event to customer with ID ${this.customerId}`);
```

File: components/customer_io/actions/send-event-to/send-event-to.mjs [hunk 4]
```diff
@@ -1,42 +1,38 @@
...
-      url: `https://track.customer.io/api/v1/customers/${this.customer_id}/events`,
-      auth: {
-        username: this.customer_io.$auth.site_id,
-        password: this.customer_io.$auth.api_key,
-      },
-      data,
-    };
-    return axios($, config);
+    const res = await this.app.sendEventTo(this.customerId, data, $);
+    $.export("$summary", "Successfully sent event");
```
Comment (lcaresia) [inline components/customer_io/actions/send-event-to/send-event-to.mjs:35]: Obs: I didn't tested this.

```suggestion
    $.export("$summary", `Successfully sent event to customer with ID ${this.customerId}`);
```

File: components/customer_io/customer_io.app.mjs [hunk 5]
```diff
@@ -1,11 +1,62 @@
...
+  },
   methods: {
-    // this.$auth contains connected account data
-    authKeys() {
-      console.log(Object.keys(this.$auth));
+    _getBaseUrl() {
+      return "https://track.customer.io/api/v1";
+    },
+    _getHeaders() {
+      const basicauthUserPwd = `${this.$auth.site_id}:${this.$auth.api_key}`;
```
Comment (lcaresia) [inline components/customer_io/customer_io.app.mjs:18]: Move `this.$auth.api_key` and `this.$auth.site_id` to a getter method.

```suggestion
      const basicauthUserPwd = `${this._siteId()}:${this._apiKey()}`;
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
Thanks for your contribution!

I requested just some fews changes!

File: components/customer_io/actions/send-event-to-customer-io/send-event-to-customer-io.mjs (hunk 3 above)
Comment (vellames) [reply components/customer_io/actions/send-event-to-customer-io/send-event-to-customer-io.mjs:42]: Not possible, nothing is returned

File: components/customer_io/actions/send-event-to/send-event-to.mjs (hunk 4 above)
Comment (vellames) [reply components/customer_io/actions/send-event-to/send-event-to.mjs:35]: Not possible, nothing is returned

File: components/customer_io/customer_io.app.mjs (hunk 5 above)
Comment (vellames) [reply components/customer_io/customer_io.app.mjs:18]: Not possible, nothing is returned

File: components/customer_io/actions/create-or-update-customer/create-or-update-customer.mjs (hunk 2 above)
Comment (vellames) [reply components/customer_io/actions/create-or-update-customer/create-or-update-customer.mjs:41]: Not possible, nothing is returned

**Comment by vellames** [comment]
Looks like we have a pipeline problem. I don't need to change the version because it was not published yet
I am abandoning this PR and creating a new one
Cc @dannyroosevelt @dylburger

**Pull request closed without merging**